File: desktop/openin.go
package main

import (
	"embed"
	"encoding/base64"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"sync"
)

//go:embed assets/apps/*.png
var openInIcons embed.FS

type OpenInTarget struct {
	ID    string `json:"id"`
	Label string `json:"label"`
	Icon  string `json:"icon"`
}

type targetDef struct {
	id       string
	label    string
	iconPng  string
	detect   func() string
	launch   func(appPath, projectPath string) error
	openFile func(appPath, filePath string, line, col int) error
}

func iconDataURI(name string) string {
	data, err := openInIcons.ReadFile("assets/apps/" + name)
	if err != nil {
		return ""
	}
	return "data:image/png;base64," + base64.StdEncoding.EncodeToString(data)
}

func appDirs() []string {
	home, _ := os.UserHomeDir()
	return []string{"/Applications", filepath.Join(home, "Applications")}
}

// expandAppCandidates maps /Applications/X.app → [/Applications/X.app, ~/Applications/X.app]
// so user-installed apps are checked alongside system-wide ones.
func expandAppCandidates(path string) []string {
	if strings.HasPrefix(path, "/Applications/") {
		home, _ := os.UserHomeDir()
		return []string{path, filepath.Join(home, "Applications", path[len("/Applications/"):])}
	}
	return []string{path}
}

func detectByPaths(paths ...string) string {
	for _, p := range paths {
		for _, cand := range expandAppCandidates(p) {
			if _, err := os.Stat(cand); err == nil {
				return cand
			}
		}
	}
	return ""
}

// detectByPrefix is a fallback for apps with variant bundle names (e.g. "Cursor Nightly.app").
func detectByPrefix(prefix string) string {
	needle := strings.ToLower(prefix)
	for _, dir := range appDirs() {
		entries, err := os.ReadDir(dir)
		if err != nil {
			continue
		}
		for _, e := range entries {
			n := strings.ToLower(e.Name())
			if strings.HasPrefix(n, needle) && strings.HasSuffix(n, ".app") {
				return filepath.Join(dir, e.Name())
			}
		}
	}
	return ""
}

func launchOpenA(appName, projectPath string) error {
	return exec.Command("open", "-a", appName, projectPath).Run()
}

func launchAppleScript(script string) error {
	return exec.Command("osascript", "-e", script).Run()
}

// appleScriptEscape escapes a string for embedding inside an AppleScript double-quoted literal.
func appleScriptEscape(s string) string {
	return strings.ReplaceAll(strings.ReplaceAll(s, `\`, `\\`), `"`, `\"`)
}

func launchTerminalApp(_ string, projectPath string) error {
	script := fmt.Sprintf(`tell application "Terminal" to do script "cd %s; clear"`,
		appleScriptEscape(projectPath))
	if err := launchAppleScript(script); err != nil {
		return err
	}
	return launchAppleScript(`tell application "Terminal" to activate`)
}

func launchITerm(_ string, projectPath string) error {
	script := fmt.Sprintf(`tell application "iTerm"
  activate
  create window with default profile
  tell current session of current window to write text "cd %s; clear"
end tell`, appleScriptEscape(projectPath))
	return launchAppleScript(script)
}

// formatPathSpec produces "path", "path:line", or "path:line:col" depending on inputs.
// Most editors with a CLI accept this form (VS Code/Cursor/Windsurf via -g, Sublime/Zed bare).
func formatPathSpec(path string, line, col int) string {
	if line <= 0 {
		return path
	}
	if col <= 0 {
		return fmt.Sprintf("%s:%d", path, line)
	}
	return fmt.Sprintf("%s:%d:%d", path, line, col)
}

// vscodeFamilyOpenFile returns an openFile func for VS Code-derived editors that ship
// a CLI binary at <App>/Contents/Resources/app/bin/<binName> and accept `-g path:line:col`.
func vscodeFamilyOpenFile(binName string) func(string, string, int, int) error {
	return func(appPath, filePath string, line, col int) error {
		bin := filepath.Join(appPath, "Contents", "Resources", "app", "bin", binName)
		return exec.Command(bin, "-g", formatPathSpec(filePath, line, col)).Run()
	}
}

func sublimeOpenFile(appPath, filePath string, line, col int) error {
	bin := filepath.Join(appPath, "Contents", "SharedSupport", "bin", "subl")
	return exec.Command(bin, formatPathSpec(filePath, line, col)).Run()
}

func zedOpenFile(appPath, filePath string, line, col int) error {
	bin := filepath.Join(appPath, "Contents", "MacOS", "cli")
	return exec.Command(bin, formatPathSpec(filePath, line, col)).Run()
}

func launchGhostty(_ string, projectPath string) error {
	shell := os.Getenv("SHELL")
	if shell == "" {
		shell = "/bin/zsh"
	}
	return exec.Command("open", "-na", "Ghostty.app", "--args",
		"-e", shell, "-lc", fmt.Sprintf("cd %q && exec %s", projectPath, shell)).Run()
}

// Registry order is the display order.
var targets = []targetDef{
	{
		id: "cursor", label: "Cursor", iconPng: "cursor.png",
		detect: func() string {
			if p := detectByPaths(
				"/Applications/Cursor.app",
				"/Applications/Cursor Nightly.app",
			); p != "" {
				return p
			}
			return detectByPrefix("Cursor")
		},
		launch:   func(_, path string) error { return launchOpenA("Cursor", path) },
		openFile: vscodeFamilyOpenFile("cursor"),
	},
	{
		id: "vscode", label: "VS Code", iconPng: "vscode.png",
		detect: func() string {
			return detectByPaths(
				"/Applications/Visual Studio Code.app",
				"/Applications/Code.app",
			)
		},
		launch:   func(_, path string) error { return launchOpenA("Visual Studio Code", path) },
		openFile: vscodeFamilyOpenFile("code"),
	},
	{
		id: "vscode-insiders", label: "VS Code Insiders", iconPng: "vscode-insiders.png",
		detect: func() string {
			return detectByPaths(
				"/Applications/Visual Studio Code - Insiders.app",
				"/Applications/Code - Insiders.app",
			)
		},
		launch:   func(_, path string) error { return launchOpenA("Visual Studio Code - Insiders", path) },
		openFile: vscodeFamilyOpenFile("code-insiders"),
	},
	{
		id: "windsurf", label: "Windsurf", iconPng: "windsurf.png",
		detect:   func() string { return detectByPaths("/Applications/Windsurf.app") },
		launch:   func(_, path string) error { return launchOpenA("Windsurf", path) },
		openFile: vscodeFamilyOpenFile("windsurf"),
	},
	{
		id: "zed", label: "Zed", iconPng: "zed.png",
		detect:   func() string { return detectByPaths("/Applications/Zed.app", "/Applications/Zed Preview.app") },
		launch:   func(_, path string) error { return launchOpenA("Zed", path) },
		openFile: zedOpenFile,
	},
	{
		id: "xcode", label: "Xcode", iconPng: "xcode.png",
		detect: func() string { return detectByPaths("/Applications/Xcode.app") },
		launch: func(_, path string) error { return launchOpenA("Xcode", path) },
	},
	{
		id: "sublime-text", label: "Sublime Text", iconPng: "sublime-text.png",
		detect:   func() string { return detectByPaths("/Applications/Sublime Text.app") },
		launch:   func(_, path string) error { return launchOpenA("Sublime Text", path) },
		openFile: sublimeOpenFile,
	},
	{
		id: "terminal", label: "Terminal", iconPng: "terminal.png",
		detect: func() string {
			return detectByPaths("/System/Applications/Utilities/Terminal.app", "/Applications/Utilities/Terminal.app")
		},
		launch: launchTerminalApp,
	},
	{
		id: "iterm2", label: "iTerm2", iconPng: "iterm2.png",
		detect: func() string { return detectByPaths("/Applications/iTerm.app", "/Applications/iTerm2.app") },
		launch: launchITerm,
	},
	{
		id: "ghostty", label: "Ghostty", iconPng: "ghostty.png",
		detect: func() string { return detectByPaths("/Applications/Ghostty.app") },
		launch: launchGhostty,
	},
	{
		id: "warp", label: "Warp", iconPng: "warp.png",
		detect: func() string { return detectByPaths("/Applications/Warp.app") },
		launch: func(_, path string) error { return launchOpenA("Warp", path) },
	},
	{
		id: "finder", label: "Finder", iconPng: "finder.png",
		detect: func() string { return "/System/Library/CoreServices/Finder.app" },
		launch: func(_, path string) error { return exec.Command("open", path).Run() },
	},
}

var targetsByID = func() map[string]*targetDef {
	m := make(map[string]*targetDef, len(targets))
	for i := range targets {
		m[targets[i].id] = &targets[i]
	}
	return m
}()

// Detection runs once per session — new app installs require an app restart.
var (
	listCacheMu sync.Mutex
	listCache   []OpenInTarget
)

func (a *App) ListOpenInTargets() []OpenInTarget {
	listCacheMu.Lock()
	defer listCacheMu.Unlock()
	if listCache != nil {
		return listCache
	}

	out := make([]OpenInTarget, 0, len(targets))
	for _, t := range targets {
		if t.detect() == "" {
			continue
		}
		out = append(out, OpenInTarget{
			ID:    t.id,
			Label: t.label,
			Icon:  iconDataURI(t.iconPng),
		})
	}
	listCache = out
	return out
}

func (a *App) OpenIn(targetID, projectPath string) error {
	t, ok := targetsByID[targetID]
	if !ok {
		return fmt.Errorf("unknown open-in target: %s", targetID)
	}
	appPath := t.detect()
	if appPath == "" {
		return fmt.Errorf("%s is not installed", t.label)
	}
	if projectPath == "" {
		return fmt.Errorf("empty project path")
	}
	if strings.HasPrefix(projectPath, "~/") {
		home, _ := os.UserHomeDir()
		projectPath = filepath.Join(home, projectPath[2:])
	}
	return t.launch(appPath, projectPath)
}

// FileExists reports whether absPath points to a regular file (not a directory).
// Used by the terminal link provider to filter out false-positive path matches
// before underlining them.
func (a *App) FileExists(absPath string) bool {
	if absPath == "" {
		return false
	}
	info, err := os.Stat(absPath)
	if err != nil {
		return false
	}
	return !info.IsDir()
}

// Cap to keep the renderer responsive — anything larger than this and the
// "open externally" path is the right answer anyway.
const readFileMaxBytes = 5 * 1024 * 1024 // 5 MiB

// ReadFile returns the contents of absPath. Errors out if the file is too
// large or doesn't exist; in those cases the modal can render a placeholder
// and prompt the user to open it externally.
func (a *App) ReadFile(absPath string) (string, error) {
	if absPath == "" {
		return "", fmt.Errorf("empty file path")
	}
	data, err := os.ReadFile(absPath)
	if err != nil {
		return "", err
	}
	if int64(len(data)) > readFileMaxBytes {
		return "", fmt.Errorf("file too large to preview (%d bytes)", len(data))
	}
	return string(data), nil
}

// pickFileEditor returns the first installed target that supports openFile,
// in registry order. Used as the auto-detect fallback when no preference is set.
func pickFileEditor() *targetDef {
	for i := range targets {
		t := &targets[i]
		if t.openFile == nil {
			continue
		}
		if t.detect() != "" {
			return t
		}
	}
	return nil
}

// OpenFileInEditor opens absPath in the user's preferred editor, jumping to
// line:col when both are positive. Empty editorID auto-picks the first
// installed editor that supports file-level open. Falls back to `open path`
// when nothing better is available.
func (a *App) OpenFileInEditor(editorID, absPath string, line, col int) error {
	if absPath == "" {
		return fmt.Errorf("empty file path")
	}
	if _, err := os.Stat(absPath); err != nil {
		return fmt.Errorf("file not found: %s", absPath)
	}

	if editorID != "" {
		t, ok := targetsByID[editorID]
		if !ok {
			return fmt.Errorf("unknown editor: %s", editorID)
		}
		appPath := t.detect()
		if appPath == "" {
			return fmt.Errorf("%s is not installed", t.label)
		}
		if t.openFile != nil {
			return t.openFile(appPath, absPath, line, col)
		}
		return exec.Command("open", "-a", t.label, absPath).Run()
	}

	if t := pickFileEditor(); t != nil {
		return t.openFile(t.detect(), absPath, line, col)
	}
	return exec.Command("open", absPath).Run()
}

File: desktop/frontend/src-tauri/src/openin.rs
// "Open in <editor/app>" — port of desktop/openin.go. macOS-only. App icons are
// embedded from assets/apps/*.png and returned as data: URIs.
use crate::config::expand_home;
use crate::files::resolve_existing_file;
use base64::Engine;
use serde::Serialize;
use std::process::Command;

#[derive(Serialize)]
pub struct OpenInTarget {
    pub id: String,
    pub label: String,
    pub icon: String,
    #[serde(rename = "fileOnly", skip_serializing_if = "std::ops::Not::not")]
    pub file_only: bool,
}

struct Target {
    id: &'static str,
    label: &'static str,
    icon: &'static str, // png filename, or "" when no asset
    file_only: bool,
}

// Display order = this order.
const TARGETS: &[Target] = &[
    Target { id: "cursor", label: "Cursor", icon: "cursor.png", file_only: false },
    Target { id: "vscode", label: "Visual Studio Code", icon: "vscode.png", file_only: false },
    Target { id: "vscode-insiders", label: "Visual Studio Code - Insiders", icon: "vscode-insiders.png", file_only: false },
    Target { id: "windsurf", label: "Windsurf", icon: "windsurf.png", file_only: false },
    Target { id: "zed", label: "Zed", icon: "zed.png", file_only: false },
    Target { id: "xcode", label: "Xcode", icon: "xcode.png", file_only: false },
    Target { id: "sublime-text", label: "Sublime Text", icon: "sublime-text.png", file_only: false },
    Target { id: "webstorm", label: "WebStorm", icon: "", file_only: false },
    Target { id: "typora", label: "Typora", icon: "typora.png", file_only: true },
    Target { id: "terminal", label: "Terminal", icon: "terminal.png", file_only: false },
    Target { id: "iterm2", label: "iTerm", icon: "iterm2.png", file_only: false },
    Target { id: "ghostty", label: "Ghostty", icon: "ghostty.png", file_only: false },
    Target { id: "warp", label: "Warp", icon: "warp.png", file_only: false },
    Target { id: "finder", label: "Finder", icon: "finder.png", file_only: false },
];

fn home() -> String {
    dirs::home_dir().unwrap_or_default().to_string_lossy().into_owned()
}

/// `/Applications/X.app` also checks `~/Applications/X.app`.
fn app_candidates(path: &str) -> Vec<String> {
    match path.strip_prefix("/Applications/") {
        Some(rest) => vec![path.to_string(), format!("{}/Applications/{}", home(), rest)],
        None => vec![path.to_string()],
    }
}

pub(crate) fn detect_by_paths(paths: &[&str]) -> Option<String> {
    for p in paths {
        for cand in app_candidates(p) {
            if std::fs::metadata(&cand).is_ok() {
                return Some(cand);
            }
        }
    }
    None
}

fn detect_by_prefix(prefix: &str) -> Option<String> {
    let pl = prefix.to_lowercase();
    for dir in ["/Applications".to_string(), format!("{}/Applications", home())] {
        if let Ok(rd) = std::fs::read_dir(&dir) {
            for e in rd.flatten() {
                let name = e.file_name().to_string_lossy().to_lowercase();
                if name.starts_with(&pl) && name.ends_with(".app") {
                    return Some(e.path().to_string_lossy().into_owned());
                }
            }
        }
    }
    None
}

/// Detected app bundle path, or None when not installed.
fn detect(id: &str) -> Option<String> {
    match id {
        "cursor" => detect_by_paths(&["/Applications/Cursor.app", "/Applications/Cursor Nightly.app"])
            .or_else(|| detect_by_prefix("Cursor")),
        "vscode" => detect_by_paths(&["/Applications/Visual Studio Code.app", "/Applications/Code.app"]),
        "vscode-insiders" => detect_by_paths(&[
            "/Applications/Visual Studio Code - Insiders.app",
            "/Applications/Code - Insiders.app",
        ]),
        "windsurf" => detect_by_paths(&["/Applications/Windsurf.app"]),
        "zed" => detect_by_paths(&["/Applications/Zed.app", "/Applications/Zed Preview.app"]),
        "xcode" => detect_by_paths(&["/Applications/Xcode.app"]),
        "sublime-text" => detect_by_paths(&["/Applications/Sublime Text.app"]),
        "webstorm" => detect_by_paths(&["/Applications/WebStorm.app"]).or_else(|| detect_by_prefix("WebStorm")),
        "typora" => detect_by_paths(&["/Applications/Typora.app"]),
        "terminal" => detect_by_paths(&[
            "/System/Applications/Utilities/Terminal.app",
            "/Applications/Utilities/Terminal.app",
        ]),
        "iterm2" => detect_by_paths(&["/Applications/iTerm.app", "/Applications/iTerm2.app"]),
        "ghostty" => detect_by_paths(&["/Applications/Ghostty.app"]),
        "warp" => detect_by_paths(&["/Applications/Warp.app"]),
        "finder" => Some("/System/Library/CoreServices/Finder.app".to_string()),
        _ => None,
    }
}

fn icon_data_uri(file: &str) -> String {
    let bytes: &[u8] = match file {
        "cursor.png" => include_bytes!("../assets/apps/cursor.png"),
        "vscode.png" => include_bytes!("../assets/apps/vscode.png"),
        "vscode-insiders.png" => include_bytes!("../assets/apps/vscode-insiders.png"),
        "windsurf.png" => include_bytes!("../assets/apps/windsurf.png"),
        "zed.png" => include_bytes!("../assets/apps/zed.png"),
        "xcode.png" => include_bytes!("../assets/apps/xcode.png"),
        "sublime-text.png" => include_bytes!("../assets/apps/sublime-text.png"),
        "typora.png" => include_bytes!("../assets/apps/typora.png"),
        "terminal.png" => include_bytes!("../assets/apps/terminal.png"),
        "iterm2.png" => include_bytes!("../assets/apps/iterm2.png"),
        "ghostty.png" => include_bytes!("../assets/apps/ghostty.png"),
        "warp.png" => include_bytes!("../assets/apps/warp.png"),
        "finder.png" => include_bytes!("../assets/apps/finder.png"),
        _ => return String::new(),
    };
    format!(
        "data:image/png;base64,{}",
        base64::engine::general_purpose::STANDARD.encode(bytes)
    )
}

fn target(id: &str) -> Option<&'static Target> {
    TARGETS.iter().find(|t| t.id == id)
}

fn shell_quote(s: &str) -> String {
    format!("'{}'", s.replace('\'', "'\\''"))
}

fn applescript_escape(s: &str) -> String {
    s.replace('\\', "\\\\").replace('"', "\\\"")
}

fn run(c: &mut Command) -> Result<(), String> {
    let status = c.status().map_err(|e| e.to_string())?;
    if !status.success() {
        return Err("command failed".into());
    }
    Ok(())
}

// ---- commands ---------------------------------------------------------------

#[tauri::command(async)]
pub fn list_open_in_targets() -> Vec<OpenInTarget> {
    TARGETS
        .iter()
        .filter(|t| detect(t.id).is_some())
        .map(|t| OpenInTarget {
            id: t.id.into(),
            label: t.label.into(),
            icon: if t.icon.is_empty() {
                String::new()
            } else {
                icon_data_uri(t.icon)
            },
            file_only: t.file_only,
        })
        .collect()
}

#[tauri::command(async)]
pub fn open_in(target_id: String, project_path: String) -> Result<(), String> {
    let t = target(&target_id).ok_or_else(|| format!("unknown open-in target: {target_id}"))?;
    detect(&target_id).ok_or_else(|| format!("{} is not installed", t.label))?;
    if project_path.is_empty() {
        return Err("empty project path".into());
    }
    let path = expand_home(&project_path);
    match target_id.as_str() {
        "finder" => run(Command::new("open").arg(&path)),
        "terminal" => launch_terminal(&path),
        "iterm2" => launch_iterm(&path),
        "ghostty" => launch_ghostty(&path),
        _ => run(Command::new("open").args(["-a", t.label, &path])),
    }
}

#[tauri::command(async)]
pub fn open_file_in_editor(
    editor_id: String,
    abs_path: String,
    line: i64,
    col: i64,
) -> Result<(), String> {
    let abs = resolve_existing_file(&abs_path)?;
    if !editor_id.is_empty() {
        let t = target(&editor_id).ok_or_else(|| format!("unknown editor: {editor_id}"))?;
        let app_path = detect(&editor_id).ok_or_else(|| format!("{} is not installed", t.label))?;
        return open_file_with(&editor_id, t.label, &app_path, &abs, line, col);
    }
    // No editor specified: first installed target with a file-open recipe.
    for t in TARGETS {
        if editor_has_recipe(t.id) {
            if let Some(app_path) = detect(t.id) {
                return open_file_with(t.id, t.label, &app_path, &abs, line, col);
            }
        }
    }
    run(Command::new("open").arg(&abs))
}

fn editor_has_recipe(id: &str) -> bool {
    matches!(
        id,
        "cursor" | "vscode" | "vscode-insiders" | "windsurf" | "sublime-text" | "webstorm" | "zed"
    )
}

fn format_path_spec(path: &str, line: i64, col: i64) -> String {
    if line <= 0 {
        path.to_string()
    } else if col <= 0 {
        format!("{path}:{line}")
    } else {
        format!("{path}:{line}:{col}")
    }
}

fn open_file_with(
    id: &str,
    label: &str,
    app_path: &str,
    abs: &str,
    line: i64,
    col: i64,
) -> Result<(), String> {
    let spec = format_path_spec(abs, line, col);
    match id {
        "cursor" => run(Command::new(format!("{app_path}/Contents/Resources/app/bin/cursor")).args(["-g", &spec])),
        "vscode" => run(Command::new(format!("{app_path}/Contents/Resources/app/bin/code")).args(["-g", &spec])),
        "vscode-insiders" => run(Command::new(format!("{app_path}/Contents/Resources/app/bin/code-insiders")).args(["-g", &spec])),
        "windsurf" => run(Command::new(format!("{app_path}/Contents/Resources/app/bin/windsurf")).args(["-g", &spec])),
        "sublime-text" => run(Command::new(format!("{app_path}/Contents/SharedSupport/bin/subl")).arg(&spec)),
        "zed" => run(Command::new(format!("{app_path}/Contents/MacOS/cli")).arg(&spec)),
        "webstorm" => {
            let mut c = Command::new(format!("{app_path}/Contents/MacOS/webstorm"));
            if line > 0 {
                c.arg("--line").arg(line.to_string());
                if col > 0 {
                    c.arg("--column").arg(col.to_string());
                }
            }
            c.arg(abs);
            run(&mut c)
        }
        // xcode, typora, terminals: no per-line recipe — open the app on the file.
        _ => run(Command::new("open").args(["-a", label, abs])),
    }
}

fn launch_terminal(path: &str) -> Result<(), String> {
    let esc = applescript_escape(path);
    run(Command::new("osascript").args([
        "-e",
        &format!("tell application \"Terminal\" to do script \"cd {esc}; clear\""),
    ]))?;
    run(Command::new("osascript").args(["-e", "tell application \"Terminal\" to activate"]))
}

fn launch_iterm(path: &str) -> Result<(), String> {
    let esc = applescript_escape(path);
    let script = format!(
        "tell application \"iTerm\"\n  activate\n  create window with default profile\n  tell current session of current window to write text \"cd {esc}; clear\"\nend tell"
    );
    run(Command::new("osascript").args(["-e", &script]))
}

fn launch_ghostty(path: &str) -> Result<(), String> {
    let shell = std::env::var("SHELL").unwrap_or_else(|_| "/bin/zsh".into());
    let inner = format!("cd {} && exec {shell}", shell_quote(path));
    run(Command::new("open").args(["-na", "Ghostty.app", "--args", "-e", &shell, "-lc", &inner]))
}
